### PR TITLE
API to generate rST docs for configurable options

### PR DESCRIPTION
**Pull request by takluyver**
Corresponding PR to ipython/ipython coming up.

I discussed this yesterday with @minrk , and I think that each config option should only be described once, on the most general class that it applies to, and neither the config documentation nor the generated default config file should try to describe inheritance relationships. To this end, there are a couple of new methods:

- `Application._classes_inc_parents()` walks through the list of configurable classes for the application, yielding them along with all configurable parents.
- `HasTraits.class_own_traits()` is like `.class_traits()`, but only takes the traits defined on this class, not on its parents.

Also gets rid of help_classes and config_classes (#20).

### CODE REVIEW

**Comment by Carreau** [comment]
That will be of interest for @jhamrick and nbgrader.

File: traitlets/config/application.py [hunk 1]
```diff
@@ -362,6 +367,17 @@ def print_help(self, classes=False):
...
+    def document_config_options(self):
+        """Generate rST format documentation for the config options this application
+
+        Returns a multiline string.
+        """
+        chunks = []
+        for cls in self._classes_inc_parents():
+            chunks.append(cls.class_config_rst_doc())
+
+        return '\n'.join(chunks)
```
Comment (Carreau) [inline traitlets/config/application.py]: I'm sad that `_classes_inc_parents` is a generator and that you for loop and not `map(cls, lambda cls:cls.class_config_rst_doc)`. I'm sure there is even a way to get rid of the lambda.

File: traitlets/config/application.py [hunk 2]
```diff
@@ -128,15 +128,20 @@ class Application(SingletonConfigurable):
...
-        return getattr(self, 'config_classes', self.classes)
+
+    def _classes_inc_parents(self):
+        """Iterate through configurable classes, including configurable parents
+
+        Children should always be after parents, and each class should only be
+        yielded once.
+        """
+        seen = set()
+        for c in self.classes:
```
Comment (Carreau) [inline traitlets/config/application.py]: Want to make sure this is sorted to have doc order stability ?

File: traitlets/config/application.py (hunk 1 above)
Comment (takluyver) [inline traitlets/config/application.py]: I could do a genexp here, would that make you happier?

File: traitlets/config/application.py (hunk 2 above)
Comment (takluyver) [inline traitlets/config/application.py]: classes is a list, and I think we do use the order already in that list to group things meaningfully.

File: traitlets/config/application.py (hunk 1 above)
Comment (Carreau) [inline traitlets/config/application.py]: yield from all the things.

File: traitlets/config/application.py (hunk 2 above)
Comment (Carreau) [inline traitlets/config/application.py]: Ah, yes indeed. fine then.

**Comment by takluyver** [comment]
Made the code slightly briefer by using a genexp and a dict comprehension.

**Comment by Carreau** [comment]
+1

**Pull request merged**